symfony/framework-bundle/5.2/public/index.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+use App\Kernel;
+use Symfony\Component\HttpFoundation\Request;
+
+require_once ($_SERVER['APP_VENDOR_DIR'] = dirname(__DIR__).'/vendor').'/symfony/runtime/autoload.php';
+
+return function (array $context) {
+    if ($trustedProxies = $context['TRUSTED_PROXIES'] ?? false) {
+        Request::setTrustedProxies(explode(',', $trustedProxies), Request::HEADER_X_FORWARDED_ALL ^ Request::HEADER_X_FORWARDED_HOST);
+    }
+
+    if ($trustedHosts = $context['TRUSTED_HOSTS'] ?? false) {
+        Request::setTrustedHosts([$trustedHosts]);
+    }
+
+    return new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
```

Comment:
I’m still learning about the runtime component. Is it the return value of the closure that dictates which runtime is used?

Maybe a link to some runtime Part of the docs? The new runtime stuff is cool, but untraceable (and still a mostly unfamiliar concept in our world). If I wanted to change the behavior, I’d like some way to find how to do that.

Related questions:

* where is the Debug::enable() - is that in the runtime and is always enabled for a certain environment? How could I control that?

What if I need to run some code *right* after each Request object is created, but before it’s handled? We have some complex, hacky code that does that related to trusted proxies.

Thanks!

Reply:
There is always only one runtime; and it knows how to deal with a range of possible return values. You can check the [`SymfonyRuntime::start()`](https://github.com/symfony/symfony/pull/36652/files#diff-18578a7bf9e590e1557d876ecf806063R100) method to know which ones are supported right now.

If you want to change the behavior, either it's related to some global state and you might want to write your own runtime or it's something that should fit right into the close. The runtime should contain as little logic as possible. Its role is to provide a consistent runtime environement by configuring the appropriate global state.

> where is the Debug::enable

It's part of this consistent runtime environment configured by the SymfonyRuntime. You can't control it unless you configure/decorate/write the runtime. If there is something that should be commonly configured, the `$_SERVER['APP_RUNTIME_OPTIONS']` is here to help (the only key supported is `dotenv` right now, see [this discussion](https://github.com/symfony/symfony/pull/36652#discussion_r419387901) where this has been asked for, see [this one](https://github.com/symfony/symfony/pull/36652#discussion_r420604922) about the error handler.)

> What if I need to run some code right after each Request object is created, but before it’s handled? We have some complex, hacky code that does that related to trusted proxies.

I think the best way would be to decorate the `http_kernel` service. Or you override the `handle()` method in your `App\Kernel`. Or you decorate your `App\Kernel` with an anonymous class in the returned closure. etc. :)